src/core/ext/transport/chttp2/transport/writing.cc

```diff
@@ -710,17 +711,40 @@ void grpc_chttp2_end_write(grpc_chttp2_transport* t, grpc_error_handle error) {
       t->keepalive_timeout != grpc_core::Duration::Infinity()) {
     // Set ping timeout after finishing write so we don't measure our own send
     // time.
+    const auto timeout = grpc_core::IsSeparatePingFromKeepaliveEnabled()
+                             ? t->ping_timeout
+                             : t->keepalive_timeout;
     auto id = t->ping_callbacks.OnPingTimeout(
-        t->keepalive_timeout, t->event_engine.get(), [t = t->Ref()] {
+        timeout, t->event_engine.get(), [t = t->Ref()] {
           grpc_core::ApplicationCallbackExecCtx callback_exec_ctx;
           grpc_core::ExecCtx exec_ctx;
           grpc_chttp2_ping_timeout(t);
         });
     if (GRPC_TRACE_FLAG_ENABLED(grpc_ping_trace) && id.has_value()) {
       gpr_log(GPR_INFO,
               "%s[%p]: Set ping timeout timer of %s for ping id %" PRIx64,
-              t->is_client ? "CLIENT" : "SERVER", t,
-              t->keepalive_timeout.ToString().c_str(), id.value());
+              t->is_client ? "CLIENT" : "SERVER", t, timeout.ToString().c_str(),
+              id.value());
+    }
+
+    if (grpc_core::IsSeparatePingFromKeepaliveEnabled() &&
+        t->keepalive_incoming_data_wanted &&
```

Comment:
isn't keepalive_ping_timeout_handle enough of a check?

Reply:
can get reads through during the write, and i think that's sufficient to prove liveness

Reply:
ah, I misread the code